src/store/index-level.ts

```diff
@@ -37,75 +33,124 @@ export class IndexLevel {
           }
         } else {
-          checkMatches(propertyName, this.findRangeMatches(propertyName, propertyFilter, options));
+          // `propertyFilter` is a `RangeFilter`
+          const rangeMatchesPromise = this.findRangeMatches(propertyName, propertyFilter, options);
+          propertyNameToPromises[propertyName] = [rangeMatchesPromise];
         }
       } else {
-        checkMatches(propertyName, this.findExactMatches(propertyName, propertyFilter, options));
+        // propertyFilter is an EqualFilter, meaning it is a non-object primitive type
+        const exactMatchesPromise = this.findExactMatches(propertyName, propertyFilter, options);
+        propertyNameToPromises[propertyName] = [exactMatchesPromise];
       }
-
-      requiredProperties.add(propertyName);
     }
 
-    await Promise.all(promises);
+    // map of ID of all data/object -> list of missing property matches
+    // if list of missing property matches is 0, then it data/object is fully matches the filter
+    const missingPropertyMatchesForId: { [id: string]: Set<string> } = { };
+
+    // Resolve promises and find the union of results for each individual propertyName DB query
+    const matchedIDs: string[] = [ ];
+    for (const [propertyName, promises] of Object.entries(propertyNameToPromises)) {
+      // acting as an OR match for the property, any of the promises returning a match will be treated as a property match
+      for (const promise of promises) {
+        for (const id of await promise) {
+          // if first time seeing a property matching for the data/object, record all properties needing a match to track progress
+          missingPropertyMatchesForId[id] ??= new Set<string>([ ...Object.keys(filter) ]);
+
+          missingPropertyMatchesForId[id].delete(propertyName);
+          if (missingPropertyMatchesForId[id].size === 0) {
+            // full filter match, add it to return list
+            matchedIDs.push(id);
+          }
+        }
+      }
+    }
 
     return matchedIDs;
   }
 
   async delete(id: string, options?: IndexLevelOptions): Promise<void> {
-    const prefixes = await this.db.get(`__${id}__prefixes`, options);
-    if (!prefixes) {
+    const serializedIndexes = await this.db.get(`__${id}__indexes`, options);
+    if (!serializedIndexes) {
       return;
     }
 
+    const indexes = JSON.parse(serializedIndexes);
+
+    // delete all indexes associated with the data of the given ID
     const ops: LevelWrapperBatchOperation<string>[] = [ ];
-    for (const prefix of JSON.parse(prefixes)) {
-      ops.push({ type: 'del', key: this.join(prefix, id) });
+    for (const propertyName in indexes) {
+      const value = indexes[propertyName];
+      const key = this.join(propertyName, this.encodeValue(value), id);
+      ops.push({ type: 'del', key });
     }
-    ops.push({ type: 'del', key: `__${id}__prefixes` });
+
+    ops.push({ type: 'del', key: `__${id}__indexes` });
 
     return this.db.batch(ops, options);
```

Comment:
clarifying: The return type of this method is `Promise<void>` but we have this return statement here. Did you keep that for a reason?

Reply:
No, just didn't notice/bother enough. But you are right, it's redundant, went ahead and removed.